src/devices/bus/centronics/etibuffer.cpp

```diff
@@ -0,0 +1,369 @@
+	PORT_CONFNAME(0x0c, 0x0c, "Ram Size")
+//  PORT_CONFSETTING(0x00, "0K (buffer will not work, passes FF to printer)")  // ram size of 0k will pass only 0xFF on to printer, use for ram test failure
+	PORT_CONFSETTING(0x04, "16K")
+	PORT_CONFSETTING(0x08, "32K")
+	PORT_CONFSETTING(0x0c, "48K")
+
+INPUT_PORTS_END
+
+DEFINE_DEVICE_TYPE(ETIPRINTBUFFER, etiprintbuffer_device, "etiprintbuffer", "Electronics Today International Print Buffer 48K")
+
+//--------------------------------------------------
+//  device constructor
+//--------------------------------------------------
+
+
+etiprintbuffer_device::etiprintbuffer_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	etiprintbuffer_device(mconfig, ETIPRINTBUFFER, tag, owner, clock)
+{
+}
+
+etiprintbuffer_device::etiprintbuffer_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_centronics_peripheral_interface(mconfig, *this)
+	, m_maincpu(*this, "maincpu")
+	, m_screen(*this, "screen")
+	, m_ctx(*this, ETI_CENTRONICS_TAG)
+	, m_ctx_data_out(*this, "ctx_data_out")
+	, m_printerready_led(*this, "printerready_led")
+	, m_bufferready_led(*this, "bufferready_led")
+{
+}
+
+//--------------------------------------------------
+//  mem map
+//--------------------------------------------------
+
+void etiprintbuffer_device::etiprintbuffer_device_memmap(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0000, 0x7ff).rom().region("maincpu", 0);
+	map(0x1000, 0x1fff).r(FUNC(etiprintbuffer_device::eti_read_1000));
+	map(0x2000, 0x2fff).w(FUNC(etiprintbuffer_device::eti_write_2000));
+	map(0x3000, 0x3fff).w(FUNC(etiprintbuffer_device::eti_write_3000));
+}
+
+//--------------------------------------------------
+//  io map
+//--------------------------------------------------
+
+void etiprintbuffer_device::etiprintbuffer_device_iomap(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0, 0xffff).r(FUNC(etiprintbuffer_device::eti_status_r));
+}
+
+//--------------------------------------------------
+//  memory and io functions
+//--------------------------------------------------
+
+uint8_t etiprintbuffer_device::eti_status_r(offs_t offset)
+{
+	return ((m_strobereceived ? 1 : 0 ) << 6) | // d6 is m_strobe status
+			((m_busy          ? 0 : 1 ) << 7);  // d7 is not busy output
+}
+
+uint8_t etiprintbuffer_device::eti_read_1000(offs_t offset)
+{
+	output_ack(0);
+	m_strobereceived = false;
+	m_bufferready_led = !m_strobereceived;
+
+	output_busy(m_strobereceived);
+	m_ack_timer->adjust(attotime::from_usec(ACK_DELAY), TIMER_ACK);
+	return m_datalatch;
+}
+
+void etiprintbuffer_device::eti_write_2000(offs_t offset, uint8_t data)
+{
+	m_ctx_data_out->write(data);
+}
+
+void etiprintbuffer_device::eti_write_3000(offs_t offset, uint8_t data)
+{
+	m_ctx->write_strobe(CLEAR_LINE);
+	m_strobeout_timer->adjust(attotime::from_usec(STROBE_DELAY), TIMER_STROBEOUT);
+}
```

Comment:
You’re going to need synchronisation barriers or we’ll end up with the same kind of issues we’ve had with some of the Apple II printer cards where signals are set from different diexec domains.

Reply:
I haven't quite wrapped my head around the synchronization stuff, should I also synchronize when the cpu does a read to get the io status?

Reply:
It actually seems to work pretty well without any synchronization, even working with the original Grappler connected to the etibuffer to the ap2000.

I tried adding some synchronization to it and it stopped working properly.

Adding this made it stop working.


WRITE_LINE_MEMBER(etiprintbuffer_device::busy_w)
{
	machine().scheduler().synchronize(timer_expired_delegate(FUNC(etiprintbuffer_device::set_centronics_output_busy), this), state ? 1 : 0);
}

void etiprintbuffer_device::set_centronics_output_busy(void *ptr, s32 param)
{
	m_busy = u8(u32(param));
	m_printerready_led = !m_busy;
}